### ISSUE TITLE
bug - set look and feel to gtk alike causes game failure

### ISSUE THREAD

**Issue by good-old-music** [triplea-game/triplea#1208]
1.9.0.0.3101 
Kernel Linux 4.6.0-1-amd64 x86_64
Debian Stretch/testing Mate desktop 1.14.1

Setting look and feel to gtk-alike stops the game from getting past the ta startup on-screen display icon. 
Uninstalling doesn't help
reverting to an older version doesn't help
deleting TripleA folders and re-installing doesn't help
Each re-install leaves me stuck on this start-up screen (I must be missing a delete-able part somewhere).

**Comment by DanVanAtta** [comment]
This might not be a new bug, but maybe.

The look and feel is going to be set as a system property, which is stored with the OS somewhere. That is going to be sticky between re-installs even.

Looks like the needed patch is to make sure that we have error handling to fix the System prop and revert to a skin that is working.

@good-old-music - a thread dump when the game is frozen on startup would be very useful. The last bullet point gives some notes on how to do so: https://github.com/triplea-game/triplea/blob/master/docs/troubleshooting.md, likely there are decent instructions on the internet as well on how to capture a Java Thread dump (looking for stack traces, a thread dump should provide one stack trace per thread, so it'll be a quite a bit of info with about 30 lines per stack trace and perhaps 10 threads or so.)

**Comment by good-old-music** [comment]
"The look and feel is going to be set as a system property, which is stored with the OS somewhere"
Ouch, looks like I'm in for a long night.
First off I get the following onscreen from a terminal while the game is running. 

triplea.engine.version.bin:1.9

(java:3628): GLib-GObject-WARNING **: cannot register existing type 'GdkScreen'

(java:3628): GLib-CRITICAL **: g_once_init_leave: assertion 'result != 0' failed

(java:3628): Gdk-CRITICAL **: gdk_screen_is_composited: assertion 'GDK_IS_SCREEN (screen)' failed

And as output text is only 8.5Kb I'll take the liberty of posting it here too - If I am wrong to do this, just say so and I won't do it again. Hopefully you will have a better idea than me where I can go next. Cheers.

triplea.engine.version.bin:1.9
2016-09-06 18:28:24
Full thread dump OpenJDK 64-Bit Server VM (25.102-b14 mixed mode):

"DestroyJavaVM" #22 prio=5 os_prio=0 tid=0x00007f016400a800 nid=0xecb waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Console std err reader" #18 daemon prio=6 os_prio=0 tid=0x00007f0120043800 nid=0xedf in Object.wait() [0x00007f01175fb000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000eba15dd8> (a java.lang.Object)
	at java.lang.Object.wait(Object.java:502)
	at games.strategy.debug.SynchedByteArrayOutputStream.readFully(SynchedByteArrayOutputStream.java:46)
	- locked <0x00000000eba15dd8> (a java.lang.Object)
	at games.strategy.debug.ThreadReader.run(ThreadReader.java:25)
	at java.lang.Thread.run(Thread.java:745)

"Console std out reader" #17 daemon prio=6 os_prio=0 tid=0x00007f0120042000 nid=0xede in Object.wait() [0x00007f01176fc000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000eba26090> (a java.lang.Object)
	at java.lang.Object.wait(Object.java:502)
	at games.strategy.debug.SynchedByteArrayOutputStream.readFully(SynchedByteArrayOutputStream.java:46)
	- locked <0x00000000eba26090> (a java.lang.Object)
	at games.strategy.debug.ThreadReader.run(ThreadReader.java:25)
	at java.lang.Thread.run(Thread.java:745)

"AWT-EventQueue-1" #15 prio=6 os_prio=0 tid=0x00007f016446d800 nid=0xedc runnable [0x00007f0117ffd000]
   java.lang.Thread.State: RUNNABLE
	at com.sun.java.swing.plaf.gtk.GTKStyle.nativeGetXThickness(Native Method)
	at com.sun.java.swing.plaf.gtk.GTKStyle.<init>(GTKStyle.java:97)
	- locked <0x00000000ebaca240> (a java.lang.Object)
	at com.sun.java.swing.plaf.gtk.GTKStyleFactory.getStyle(GTKStyleFactory.java:109)
	- locked <0x00000000ebac1760> (a com.sun.java.swing.plaf.gtk.GTKStyleFactory)
	at com.sun.java.swing.plaf.gtk.GTKLookAndFeel.initComponentDefaults(GTKLookAndFeel.java:366)
	at com.sun.java.swing.plaf.gtk.GTKLookAndFeel.getDefaults(GTKLookAndFeel.java:310)
	at javax.swing.UIManager.setLookAndFeel(UIManager.java:539)
	at javax.swing.UIManager.setLookAndFeel(UIManager.java:583)
	at games.strategy.engine.framework.lookandfeel.LookAndFeel.lambda$setupLookAndFeel$103(LookAndFeel.java:19)
	at games.strategy.engine.framework.lookandfeel.LookAndFeel$$Lambda$33/650453719.run(Unknown Source)
	at games.strategy.ui.SwingAction.invokeAndWait(SwingAction.java:61)
	at games.strategy.engine.framework.lookandfeel.LookAndFeel.setupLookAndFeel(LookAndFeel.java:17)
	at games.strategy.engine.framework.GameRunner.lambda$main$59(GameRunner.java:176)
	at games.strategy.engine.framework.GameRunner$$Lambda$27/914504136.run(Unknown Source)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at games.strategy.engine.framework.GameRunner$1.dispatchEvent(GameRunner.java:353)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

"AWT-Shutdown" #16 prio=5 os_prio=0 tid=0x00007f016446b800 nid=0xedb in Object.wait() [0x00007f011c145000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ebb59fc0> (a java.lang.Object)
	at java.lang.Object.wait(Object.java:502)
	at sun.awt.AWTAutoShutdown.run(AWTAutoShutdown.java:295)
	- locked <0x00000000ebb59fc0> (a java.lang.Object)
	at java.lang.Thread.run(Thread.java:745)

"AWT-XAWT" #13 daemon prio=6 os_prio=0 tid=0x00007f016438d800 nid=0xed8 runnable [0x00007f0127592000]
   java.lang.Thread.State: RUNNABLE
	at sun.awt.X11.XToolkit.waitForEvents(Native Method)
	at sun.awt.X11.XToolkit.run(XToolkit.java:568)
	at sun.awt.X11.XToolkit.run(XToolkit.java:532)
	at java.lang.Thread.run(Thread.java:745)

"Java2D Disposer" #11 daemon prio=10 os_prio=0 tid=0x00007f016433b000 nid=0xed7 in Object.wait() [0x00007f0127693000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000eba0afa0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000000eba0afa0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at sun.java2d.Disposer.run(Disposer.java:148)
	at java.lang.Thread.run(Thread.java:745)

"Timer-0" #9 daemon prio=5 os_prio=0 tid=0x00007f0164262000 nid=0xed6 in Object.wait() [0x00007f01545ec000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ebb67f00> (a java.util.TaskQueue)
	at java.util.TimerThread.mainLoop(Timer.java:552)
	- locked <0x00000000ebb67f00> (a java.util.TaskQueue)
	at java.util.TimerThread.run(Timer.java:505)

"Service Thread" #7 daemon prio=9 os_prio=0 tid=0x00007f0164135800 nid=0xed4 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread1" #6 daemon prio=9 os_prio=0 tid=0x00007f0164132800 nid=0xed3 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 tid=0x00007f016412f800 nid=0xed2 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x00007f016412d800 nid=0xed1 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x00007f0164107000 nid=0xed0 in Object.wait() [0x00007f0154dfc000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ebb68ae8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000000ebb68ae8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x00007f0164102000 nid=0xecf in Object.wait() [0x00007f0154efd000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ebb6cf80> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x00000000ebb6cf80> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"VM Thread" os_prio=0 tid=0x00007f01640fa800 nid=0xece runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x00007f016401f800 nid=0xecc runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x00007f0164021000 nid=0xecd runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x00007f0164144000 nid=0xed5 waiting on condition 

JNI global references: 381

Heap
 PSYoungGen      total 17920K, used 14010K [0x00000000eab00000, 0x00000000ebf00000, 0x0000000100000000)
  eden space 15360K, 74% used [0x00000000eab00000,0x00000000eb636098,0x00000000eba00000)
  from space 2560K, 98% used [0x00000000eba00000,0x00000000ebc78840,0x00000000ebc80000)
  to   space 2560K, 0% used [0x00000000ebc80000,0x00000000ebc80000,0x00000000ebf00000)
 ParOldGen       total 40960K, used 2071K [0x00000000c0000000, 0x00000000c2800000, 0x00000000eab00000)
  object space 40960K, 5% used [0x00000000c0000000,0x00000000c0205f70,0x00000000c2800000)
 Metaspace       used 18647K, capacity 18958K, committed 19200K, reserved 1067008K
  class space    used 2201K, capacity 2298K, committed 2304K, reserved 1048576K

**Comment by good-old-music** [comment]
Just in case anyone makes the same mistake, the solution for me was to find the java file that sets the preference. In Linux it was /home/user/.java/.userPrefs/games/strategy/triplea/settings/prefs.xml and the line that needs to be altered needs to made into
<entry key="LOOK_AND_FEEL_PREF" value="org.pushingpixels.substance.api.skin.SubstanceGraphiteLookAndFeel"/>

to reset it to the original value.
Cheers, and thanks for the game engine all you developers.

**Comment by DanVanAtta** [comment]
Re-opening, would like to track and fix.

The game should not crash given any user input, you should be to click or select anything, and not have to go through this level of effort to recover the game. Besides, this will happen to someone else.

**Comment by DanVanAtta** [comment]
https://github.com/triplea-game/triplea/pull/1211 Should help fix this. After that change is merged,  we will set the look and feel before setting the system property. This way if the game crashes setting the look and feel, we'll never get to the system property, and on restart you should still be on the old one.

@good-old-music , once https://github.com/triplea-game/triplea/pull/1211 is merged, would you mind verifying that is what happens?  If so, please stay tuned and we'll have a new build number.

I do not have the same look and feel on my system. There is a capability to grab "system installed" look and feels. I suspect it is one of these and it is broken..

**Comment by good-old-music** [comment]
Delighted too, I enjoy these problems (when they are recoverable) and I learnt a little too. 
Wasn't sure about best practice on closing or not.

**Comment by DanVanAtta** [comment]
@good-old-music  - update has been merged in, would you mind giving the bugged look and feel a try on the latest game engine?

**Comment by good-old-music** [comment]
Here is what occurs on setting to a known broken value. Other L&F values work normally at a quick glance.
https://youtu.be/iUOvOMivtrw

Thought a video might explain it better. Resets to default L&F on a restart after killing process. Will test more thoroughly when I get home later.

**Comment by DanVanAtta** [comment]
The reset to default L&F on restart was the thing to test - idea being at least now you don't have to go to great lengths to recover the game. If you would, could you try opening the error console before switching look and feel, there might be something printed.

**Comment by good-old-music** [comment]
Nothing in the console.
Only two are non-working and cause a lock on the game, metal and gtk, the rest all operate as expected.

Dump for gtk look and feel is available here if any use http://pastebin.com/PiwQb4em

May be easier to remove those two as game options as all the rest seem fine.

**Comment by DanVanAtta** [comment]
Some of the look and feels, metal for example, are built in. Makes it a bit more tricky to remove them from the list.

Meanwhile, adding this to the bug backlong and removing the 1.9 milestone for two reasons:
- this was a problem present in 1.8.0.9, it was made no worse
- game engine has been updated so the problem is now at least recoverable
